app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,264 @@
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int): Message {
+        val message = Message()
+        message.what = what
+        return message
+    }
+
+    private fun constructMessage(message: Message?, what: Int): Message {
+        var messageValue = message
+        if (messageValue == null) {
+            messageValue = Message()
+        }
+        messageValue.what = what
+        return messageValue
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(looper: Looper?, val connectionMonitor: ConnectionMonitor) : Handler(looper!!) {
+        override fun handleMessage(msg: Message) {
+            processMessage(msg)
+        }
+
+        fun processMessage(msg: Message) {
+            if (msg.what == MSG_ADD_ACTIVE_NETWORK) {
+                processActiveNetwork()
+            } else if (msg.what == MSG_ADD_ALL_NETWORKS) {
+                processAllNetwork()
+            }
+        }
+
+        /**
+         * Handles the active network.
+         * Get active network from connectivity-manager
+         * Set the vpn underlying network when the current active network is different from
+         * already assigned active network or if there is difference in the available network set
+         */
+        private fun processActiveNetwork(){
+            val activeNetwork: Network? = connectionMonitor.connectivityManager.activeNetwork
+            val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork?.networkHandle
+            connectionMonitor.currentActiveNetwork = activeNetwork
+            if(!isNewNetwork){
+                return
+            }
+            if (activeNetwork != null) {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+                connectionMonitor.networkListener.onUpdateActiveNetwork(activeNetwork)
+            } else {
+                connectionMonitor.networkListener.onNetworkDisconnected()
+            }
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(connectionMonitor.currentActiveNetwork)?.typeName.toString()}")
+        }
+
+
+
+        /**
+         * Adds all the available network to the underlying network.
+         */
+        private fun processAllNetwork() {
+            val newActiveNetwork = connectionMonitor.connectivityManager.activeNetwork
+            var isNewNetwork = connectionMonitor.currentActiveNetwork != newActiveNetwork
```

Comment:
why is the check here different from what's above? Which one is correct?!

here:
```
  var isNewNetwork = connectionMonitor.currentActiveNetwork != newActiveNetwork
```

above:
```
  val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork?.networkHandle
```

Reply:
Both are right. NetworkHandle will have the long value which is specific for the network. Now using connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork?.networkHandle everywhere